pyomo/contrib/trustregion/interface.py

```diff
@@ -0,0 +1,406 @@
+
+        if len(self.data.objs) != 1:
+            raise ValueError(
+                "replaceExternalFunctionsWithVariables: "
+                "TrustRegion only supports models with a single active Objective.")
+        self._remove_ef_from_expr(self.data.objs[0])
+
+        for i in self.data.ef_outputs:
+            self.data.ef_inputs[i] = \
+                list(identify_variables(
+                    self.data.truth_models[self.data.ef_outputs[i]],
+                    include_fixed=False)
+                )
+        self.data.all_variables.update(self.data.ef_outputs.values())
+        self.data.all_variables = list(self.data.all_variables)
+
+    def createConstraints(self):
+        """
+        Create the basis constraint y = b(w) (equation 3) and the
+        surrogate model constraint y = r_k(w) (equation 5)
+
+        Both constraints are immediately deactivated after creation and
+        are activated later as necessary.
+        """
+        b = self.data
+        # This implements: y = b(w) from Yoshio/Biegler (2020)
+        @b.Constraint(b.ef_outputs.index_set())
+        def basis_constraint(b, i):
+            ef_output_var = b.ef_outputs[i]
+            return ef_output_var == b.basis_expressions[ef_output_var]
+        b.basis_constraint.deactivate()
+
+        b.INPUT_OUTPUT = Set(initialize=(
+            (i, j) for i in b.ef_outputs.index_set()
+            for j in range(len(b.ef_inputs[i]))
+        ))
+        b.basis_model_output = Param(b.ef_outputs.index_set(), mutable=True)
+        b.grad_basis_model_output = Param(b.INPUT_OUTPUT, mutable=True)
+        b.truth_model_output = Param(b.ef_outputs.index_set(), mutable=True)
+        b.grad_truth_model_output = Param(b.INPUT_OUTPUT, mutable=True)
+        b.value_of_ef_inputs = Param(b.INPUT_OUTPUT, mutable=True)
+        # This implements: y = r_k(w)
+        @b.Constraint(b.ef_outputs.index_set())
+        def sm_constraint_basis(b, i):
+            ef_output_var = b.ef_outputs[i]
+            return ef_output_var == b.basis_expressions[ef_output_var] + \
+                b.truth_model_output[i] - b.basis_model_output[i] + \
+                sum((b.grad_truth_model_output[i, j]
+                      - b.grad_basis_model_output[i, j])
+                    * (w - b.value_of_ef_inputs[i, j])
+                    for j, w in enumerate(b.ef_inputs[i]))
+        b.sm_constraint_basis.deactivate()
+
+    def getCurrentDecisionVariableValues(self):
+        """
+        Return current decision variable values
+        """
+        decision_values = {}
+        for var in self.decision_variables:
+            decision_values[var.name] = value(var)
+        return decision_values
+
+    def updateDecisionVariableBounds(self, radius):
+        """
+        Update the TRSP_k decision variable bounds
+
+        This corresponds to:
+            || E^{-1} (u - u_k) || <= trust_radius
+        We omit E^{-1} because we assume that the users have correctly scaled
```

Comment:
Is this a reasonable assumption?

Reply:
We note in the `README` that we assume this has been done. We might change later so that the user can give us a scaling factor, but @jsiirola and I decided that we should just state that the user should take care of that for the time being.